### DISCUSSION TITLE
Item Search on other Catalog Items

### DISCUSSION THREAD

**Discussion by aclel** [TerriaJS/terriajs#7214] [Q&A] [answered]
I'd like to do indexed item search on Catalog Items other than Cesium3DTilesCatalogItem.

Had a crack at integrating the SearchableItemMixin with a MapboxVectorTileCatalogItem, but not having much luck.

Has anyone integrated the SearchableItemMixin into other item types? Looking at some examples will probably help. I've been trying to find examples of the feature running on one of the many TerriaMap instances (NSW, VIC, NationalMap etc), but I haven't been able to find any.

In the documentation it says there are future plans to do a GeoJSON search provider - has any progress has been made on this?

**Comment by na9da** [comment, marked as answer]
Hi @aclel 

There's [one example](https://digitaltwin.vic.gov.au/portals/44/map/?share=fc1b582c-7033-4a0a-b15a-70697e4ced4d) here in DTV (had to tweak some config to get that working).

You can open the search dialog from the expand menu here:
![image](https://github.com/TerriaJS/terriajs/assets/1430/ee0be167-f322-4947-a3ee-e85df5763991)

This particular dataset only indexes numeric fields.

We have an [opensource indexer](https://github.com/terriajs/terriajs-indexer) for 3d-tiles and kml+gltf files. You can have a look at how that's implemented. Unfortunately we haven't implemented indexing for any other data types yet. There is some notes on the [design of the index](https://github.com/TerriaJS/terriajs/blob/main/doc/connecting-to-data/item-search/indexed-item-search-provider-design-notes.md#results-data-file). It is basically a bunch of CSVs:
  - 1 results data file - maps an id property to its lat, lon position
     
![image](https://github.com/TerriaJS/terriajs/assets/1430/e36f0ca6-1269-43bf-973d-602ca7ba71d3)

  - 1 csv each for searchable properties - maps the id to prop value

  - and an indexRoot - json file describing the index structure
  
We would happily welcome any contributions if you intend to develop something opensource.

For standalone data types like GeoJSON, since the whole data is available in memory, we could probably implement a new [ItemSearchProvider](https://github.com/TerriaJS/terriajs/blob/1d46281e72dbd6b88bad7f5b6d2caa34f79958ea/lib/Models/ItemSearchProviders/ItemSearchProvider.ts#L45) type that does brute force search instead of indexing data ahead-of-time.

WFS would be another ItemSearchProvider that would call the remote API instead.

**Reply by aclel** [reply]
Thank you @na9da. That example is really impressive. I'll see how I go, I may end up writing another search provider and if I do I'll contribute it back.
